### ISSUE TITLE
Detection Centroid Distances does not work across detection types (Detections vs Cells)

### ISSUE THREAD

**Issue by lacan** [qupath/qupath#1550]
## Bug report

**Describe the bug**
If we have detections and cell objects and try to run Detection Centroid Distances 2D, it will only output a result for the cells and not the detections. 

**To Reproduce**
1. Annotate a small region and use cell detection to create cells (with expansion) and classify them as "Stroma"
2. Annotate another region (It can overlap with the other, but it does not matter) and detect cells **without** expansion, label them as "Other"
3. Use **Detection Centroid Distances 2D** with the defaults
4. If you look at the Cell objects, they have a **Distance to detection with Stroma um**
5. If you look at the Detections (not expanded cells) there is no measurement

If you repeat the steps above but make only **Cell objects** or **Detections without expansion** (Where both are of the same type) then it works as expected, meaning both PathObjects have the distance measurements

**Expected behavior**
I would expect the Detection Centroid Distances 2D would work for heterogeneous groups (cells and detections) as cells are detections too.

**Desktop (please complete the following information):**
 - OS: Win 11
 - QuPath Version: 0.5.1

**Comment by petebankhead** [comment]
From memory, I think the reason it prefers cells *when present* is so that it does something sensible in the (common?) scenario that someone is using subcellular detections.

I agree that this behavior isn't obvious (and should at least be documented). I'm interested to better understand the use case where heterogenous groups should be treated the same.

**Comment by MichaelSNelson** [comment]
I agree with @petebankhead here, since subcellular detections would too often get in the way. I think it would make more sense to _force the type of objects you want to have interact in this way to all have the same type_, though I acknowledge users might need dig around on the forum a bit to look up the right lines of script to accomplish this. 

I can definitely see situations where different types of cells require different types of detection methods, for example larger macrophages that can be multinucleated requiring something like CellPose to create the boundary, which may, at the time, be simply a detection as there isn't a convenient way to create a multinucleated object. I still feel like it makes more sense to have a more convenient way to force detections to be cells, rather than trying to figure out a way to exclude subcellular detections when you do not want those interactions calculated - aside from removing the measurements after the fact.

**Comment by lacan** [comment]
Perhaps something making use of the hierarchy? The Detection Centroid Distances should apply to all detections **except for detections that are child objects of other detections**. 
This would exclude subcellular detections from the run, and would keep the logic working. However at that point it could work for both detections and cells?

**Comment by petebankhead** [comment]
I'm not sure which approach is preferable to solve this. I have the feeling there will always be a need for an alternative to whichever default approach we take.

As usual, the workaround would be a script. Here's one that shows how to get much more control over which distances are calculated, here using all detections without a parent that is a detection:

```groovy
def detections = getDetectionObjects().findAll {!it.getParent()?.isDetection()}

def stroma = detections.findAll {it.getPathClass() == getPathClass('Stroma')}
def other = detections.findAll {it.getPathClass() == getPathClass('Other')}

def pixelWidth = getCurrentServer().getPixelCalibration().getPixelWidthMicrons()
def pixelHeight = getCurrentServer().getPixelCalibration().getPixelHeightMicrons()

DistanceTools.centroidToCentroidDistance2D(
    detections, other, pixelWidth, pixelHeight, "Distance to Other µm"
)
DistanceTools.centroidToCentroidDistance2D(
    detections, stroma, pixelWidth, pixelHeight, "Distance to Stroma µm"
)
```

**Comment by petebankhead** [comment]
@ofrag You can make any object the child of another object with [`addChildObject(pathObject)`](https://qupath.github.io/javadoc/docs/qupath/lib/objects/PathObject.html#addChildObject(qupath.lib.objects.PathObject)) but I'm not sure that's relevant here. At least, it's likely to complicate things more - probably better to discuss on the forum with a specific use case, as QuPath currently isn't really designed to handle nested detection objects very intelligently.
